libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/builder/LocationProvider.kt

```diff
@@ -0,0 +1,70 @@
+package com.mapbox.navigation.core.routeoptions.builder
+
+import android.location.Location
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.core.trip.session.LocationMatcherResult
+import com.mapbox.navigation.core.trip.session.LocationObserver
+import com.mapbox.navigation.core.trip.session.TripSessionLocationProvider
+import com.mapbox.navigation.utils.internal.toPoint
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.yield
+import kotlin.coroutines.resume
+
+internal interface LocationProvider {
+    suspend fun getCurrentLocation(): CurrentLocation
+}
+
+internal data class CurrentLocation(
+    val point: Point,
+    val bearing: Double?,
+    val zLevel: Int?
+)
+
+internal class LocationFromTripSessionProvider(
+    private val tripSessionLocationProvider: TripSessionLocationProvider
+) : LocationProvider {
+    override suspend fun getCurrentLocation(): CurrentLocation {
+        val currentLocation = tripSessionLocationProvider.locationMatcherResult
+        return if (currentLocation != null) {
+            currentLocation.toCurrentLocation()
+        } else {
+            waitForTheFirstLocationEvent()
+        }
+    }
+
+    private suspend fun waitForTheFirstLocationEvent(): CurrentLocation {
+        val (result, onCleanup) = suspendCancellableCoroutine<Pair<CurrentLocation, () -> Unit>>
+        { continuation ->
+            val observer = object : LocationObserver {
+                override fun onNewRawLocation(rawLocation: Location) {
+                }
+
+                override fun onNewLocationMatcherResult(
+                    locationMatcherResult: LocationMatcherResult
+                ) {
+                    continuation.resume(
+                        Pair(
+                            locationMatcherResult.toCurrentLocation(),
+                            {
+                                tripSessionLocationProvider.unregisterLocationObserver(this)
+                            }
+                        )
+                    )
+                }
+            }
+            tripSessionLocationProvider.registerLocationObserver(observer)
+            continuation.invokeOnCancellation {
+                tripSessionLocationProvider.unregisterLocationObserver(observer)
+            }
+        }
+        yield() // you can't remove listener in the listener callback
```

Comment:
I'm not sure if that workaround is robust. @mapbox/navigation-android , do you know if there is a chance that `yield` won't schedule the reset of the function to the end of the Android's main queue?

Reply:
I'm wondering is that okay to crash when a user updates observers list during an event from observer? Maybe we should iterate through a copy of observers list to let user modify the original one?
@mapbox/navigation-android , do you guys knows what the best practises around observer pattern are?

Reply:
What is the crash? We already use the copy for iteration because of `CopyOnWriteArrayList`.

Reply:
I don't remember the type of exceptions, but it crashed because I modified the listeners list from the listener, i.e. in the same time when trip session were integrating over the list. 
@Zayankovsky , thank you for the comment. I will double check. Maybe I didn't understand the reason of crash 🙂

Reply:
It's probably [ConcurrentModificationException](https://docs.oracle.com/javase/7/docs/api/java/util/ConcurrentModificationException.html). If so, we should patch `unregisterLocationObserver` to be both ConcurrentModification  and Multi Thread safe.

Reply:
@tomaszrybakiewicz in `MapboxNavigationActivity` you can see a `LocationObserver` that unregisters itself in the callback and it does not crash. Location observers are stored in a `CopyOnWriteArraySet` which prevents `ConcurrentModificationException` in this case and is also thread-safe.